src/server/oauth/token.ts

```diff
@@ -0,0 +1,337 @@
+import { fromError } from 'zod-validation-error';
+
+import { getConfig } from '../../config.js';
+import { getTokenResult } from '../../sdks/tableau-oauth/methods.js';
+import { TableauAccessToken } from '../../sdks/tableau-oauth/types.js';
+import { setLongTimeout } from '../../utils/setLongTimeout.js';
+import { generateCodeChallenge } from './generateCodeChallenge.js';
+import { AUDIENCE } from './provider.js';
+import { mcpTokenSchema } from './schemas.js';
+import { AuthorizationCode, ClientCredentials, RefreshTokenData, UserAndTokens } from './types.js';
+
+/**
+ * OAuth 2.1 Token Endpoint
+ *
+ * Exchanges authorization code for access token.
+ * Verifies PKCE code_verifier matches the original challenge.
+ * Returns JWE containing tokens for API access.
+ */
+export function token(
+  app: express.Application,
+  authorizationCodes: Map<string, AuthorizationCode>,
+  refreshTokens: Map<string, RefreshTokenData>,
+  publicKey: KeyObject,
+): void {
+  const config = getConfig();
+
+  app.post('/oauth/token', async (req, res) => {
+    const result = mcpTokenSchema.safeParse(req.body);
+
+    if (!result.success) {
+      res.status(400).json({
+        error: 'invalid_request',
+        error_description: fromError(result.error).toString(),
+      });
+      return;
+    }
+
+    const clientCredentialsResult = verifyClientCredentials({
+      required: result.data.grantType === 'client_credentials',
+      clientId: result.data.clientId,
+      clientSecret: result.data.clientSecret,
+      clientIdSecretPairs: getConfig().oauth.clientIdSecretPairs,
+      authorizationHeader: req.headers.authorization,
+    });
+
+    if (clientCredentialsResult.isErr()) {
+      res.status(401).json({
+        error: 'invalid_client',
+        error_description: clientCredentialsResult.error,
+      });
+      return;
+    }
+
+    try {
+      switch (result.data.grantType) {
+        case 'authorization_code': {
+          // Handle authorization code exchange
+          const { code, codeVerifier } = result.data;
+          const authCode = authorizationCodes.get(code);
+          if (!authCode || authCode.expiresAt < Date.now()) {
+            authorizationCodes.delete(code);
+            res.status(400).json({
+              error: 'invalid_grant',
+              error_description: 'Invalid or expired authorization code',
+            });
+            return;
+          }
+
+          // Verify PKCE
+          const challengeFromVerifier = generateCodeChallenge(codeVerifier);
+          if (challengeFromVerifier !== authCode.codeChallenge) {
+            res.status(400).json({
+              error: 'invalid_grant',
+              error_description: 'Invalid code verifier',
+            });
+            return;
+          }
+
+          // Generate tokens
+          const refreshTokenId = randomBytes(32).toString('hex');
+          const accessToken = await createAccessToken(authCode, publicKey);
+          refreshTokens.set(refreshTokenId, {
```

Comment:
Storing the refresh tokens in memory is not a long-term solution. Please see: https://docs.google.com/document/d/1NLPNt9WflU9ZeFNG8zVbBDFtl9Nn_rd2amwNrNQtZRs/edit?tab=t.0#heading=h.ro7jv4g2g9jq